ospf6d/ospf6_tlv.h

```diff
@@ -0,0 +1,207 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * OSPFv3 Type Length Value.
+ *
+ */
+
+#ifndef OSPF6_TLV_H
+#define OSPF6_TLV_H
+
+/*
+ * Generic TLV (type, length, value) macros
+ */
+struct tlv_header {
+	uint16_t type;	 /* Type of Value */
+	uint16_t length; /* Length of Value portion only, in bytes */
+};
+
+#ifdef roundup
+#define ROUNDUP(val, gran) roundup(val, gran)
+#else /* roundup */
+#define ROUNDUP(val, gran) (((val)-1 | (gran)-1) + 1)
+#endif /* roundup */
+
+#define TLV_HDR_SIZE (sizeof(struct tlv_header))
+
+#define TLV_BODY_SIZE(tlvh) (ROUNDUP(ntohs((tlvh)->length), sizeof(uint32_t)))
+
+#define TLV_SIZE(tlvh) ((uint32_t)(TLV_HDR_SIZE + TLV_BODY_SIZE(tlvh)))
+
+#define TLV_HDR_NEXT(tlvh)                                                     \
+	((struct tlv_header *)((char *)(tlvh) + TLV_SIZE(tlvh)))
+
+#define TLV_BODY(tlvh) ((char *)(tlvh) + TLV_HDR_SIZE)
+
+/*
+ * RFC 5187 - OSPFv3 Graceful Restart - Grace-LSA
+ * Graceful restart predates Extended-LSA TLVs and IANA TLV register.
+ */
+/* Grace period TLV. */
+#define TLV_GRACE_PERIOD_TYPE 1
+#define TLV_GRACE_PERIOD_LENGTH 4
+struct tlv_grace_period {
+	struct tlv_header header;
+	uint32_t interval;
+};
+
+/* Restart reason TLV. */
+#define TLV_GRACE_RESTART_REASON_TYPE 2
+#define TLV_GRACE_RESTART_REASON_LENGTH 1
+struct tlv_grace_restart_reason {
+	struct tlv_header header;
+	uint8_t reason;
+	uint8_t reserved[3];
+};
+
+
+/*
+ * OSPFv3 Extended-LSA TLV Types
+ *
+ * Ref:
+ * Internet Assigned Numbers Authority
+ * Open Shortest Path First v3 (OSPFv3) Parameters
+ * https://www.iana.org/assignments/ospfv3-parameters/ospfv3-parameters.xhtml
+ *
+ * RFC 8362 - OSPFv3 Link State Advertisement (LSA) Extensibility
+ */
+enum ospf6_extended_lsa_tlv_types {
+	OSPF6_TLV_RESERVED = 0,
+	OSPF6_TLV_ROUTER_LINK = 1,
+	OSPF6_TLV_ATTACHED_ROUTERS = 2,
+	OSPF6_TLV_INTER_AREA_PREFIX = 3,
+	OSPF6_TLV_INTER_AREA_ROUTER = 4,
+	OSPF6_TLV_EXTERNAL_PREFIX = 5,
+	OSPF6_TLV_INTRA_AREA_PREFIX = 6,
+	OSPF6_TLV_IPV6_LL_ADDR = 7,
+	OSPF6_TLV_IPV4_LL_ADDR = 8,
+	/*
+	 * when adding new TLVs,
+	 * also update tlv_size_map[OSPF6_TLV_ENUM_END]
+	 */
+	OSPF6_TLV_ENUM_END
+};
+
+#define TLV_ROUTER_LINK_TYPE 1
+#define TLV_ROUTER_LINK_LENGTH 16U /* plus Sub-TLVs */
+/* fields correspond to struct ospf6_router_lsdesc - RFC 8362 3.2 */
+struct tlv_router_link {
+	struct tlv_header header;
+	uint8_t type;
+	uint8_t reserved;
+	uint16_t metric; /* output cost */
+	uint32_t interface_id;
+	uint32_t neighbor_interface_id;
+	in_addr_t neighbor_router_id;
+};
+
+#define TLV_ATTACHED_ROUTERS_TYPE 2
+#define TLV_ATTACHED_ROUTERS_LENGTH 4U /* times adjacent neighbors */
+/* fields correspond to struct ospf6_network_lsdesc - RFC 8362 3.3 */
+struct tlv_attached_routers {
+	struct tlv_header header;
+	in_addr_t router_id;
+};
+
+#define TLV_INTER_AREA_PREFIX_TYPE 3
+#define TLV_INTER_AREA_PREFIX_LENGTH 20U /* plus Sub-TLVs */
```

Comment:
It is incorrect to define fixed lengths for TLVs containing prefixes.

Prefix lengths are variable on the wire, and that's why they're not included in the LSA structs and why LSAs have _MIN_LENGTH macros that exclude the prefix.